EuiHighlight: converted to Typescript

ffknob opened the pull request:
### Summary

Closes https://github.com/elastic/eui/issues/2666
Converted `EuiHighlight` to Typescript.

### Checklist

- [ ] Check against **all themes** for compatability in both light and dark modes
- [ ] Checked in **mobile**
- [ ] Checked in **IE11** and **Firefox**
- [ ] Props have proper **autodocs**
- [ ] Added **documentation** examples
- [ ] Added or updated **jest tests**
- [ ] Checked for **breaking changes** and labeled appropriately
- [ ] Checked for **accessibility** including keyboard-only and screenreader modes
- [ ] A [changelog](https://github.com/elastic/eui/blob/master/CHANGELOG.md) entry exists and is marked appropriately

elasticmachine commented:
Since this is a community submitted pull request, a Jenkins build has not been kicked off automatically. Can an Elastic organization member please verify the contents of this patch and then kick off a build manually?

chandlerprall commented on src/components/highlight/highlight.tsx:
```diff
@@ -0,0 +1,79 @@
+import { CommonProps } from '../common';
+import PropTypes from 'prop-types';
+
+export type EuiHighlightProps = HTMLAttributes<HTMLDivElement> &
```
The `...rest` is spread onto a div element

```suggestion
export type EuiHighlightProps = HTMLAttributes<HTMLSpanElement> &
```

chandlerprall commented on src/components/highlight/highlight.tsx:
```diff
@@ -0,0 +1,79 @@
+  );
+};
+
+EuiHighlight.propTypes = {
```
Remove these propTypes, they will be generated automatically from the TS definitions.

ffknob replied:
So the goal is to move all propTypes to the component's interface? (just to be sure before I take the other convertions)

chandlerprall commented on src/components/highlight/highlight.tsx:
```diff
@@ -0,0 +1,79 @@
+import PropTypes from 'prop-types';
+
+export type EuiHighlightProps = HTMLAttributes<HTMLDivElement> &
+  CommonProps & {
```
`children` was marked as required in the propTypes, and is a string (instead of the standard ReactNode). Adding `children: string;` to this definition will recreate this condition.

ffknob commented:
- `CHANGELOG`: added (thought code refactory was not to be put there, my mistake)
- `HTMLSpanElement`: corrected
- `children`: included in the component's Props as `string`

One doubt about `propTypes` as general (so I can take on the other convertions): should I completly get rid of it and move all attributes to the Props interface?

chandlerprall commented:
> One doubt about propTypes as general (so I can take on the other convertions): should I completly get rid of it and move all attributes to the Props interface?

Yes, the React `propTypes` duplicate the information provided by the typescript definition, so we don't want to include them in the components. However, it is useful to provide the proptypes when components are used in non-TS projects, so we generate them (based on the TS definitions) at build time. Our docs also use the resulting proptypes to populate any `props` tabs.

chandlerprall commented on src/components/highlight/highlight.tsx:
```diff
@@ -0,0 +1,75 @@
+  CommonProps & {
+    children: string;
+
+    className?: string;
```
`className` is already provided by `CommonProps` and doesn't need to be re-specified, [common.ts#L10](https://github.com/elastic/eui/blob/641732b1011a9455c70156cefb58c8c32d11feeb/src/components/common.ts#L10)

chandlerprall commented:
Thank you once again, @ffknob !
